optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/util/ShiftRosterXlsxFileIO.java

```diff
@@ -0,0 +1,94 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaweb.employeerostering.util;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Locale;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import org.apache.poi.ss.usermodel.Cell;
+import org.apache.poi.ss.usermodel.Row;
+import org.apache.poi.ss.usermodel.Sheet;
+import org.apache.poi.ss.usermodel.Workbook;
+import org.apache.poi.xssf.usermodel.XSSFWorkbook;
+import org.optaweb.employeerostering.domain.employee.Employee;
+import org.optaweb.employeerostering.domain.roster.view.ShiftRosterView;
+import org.optaweb.employeerostering.domain.shift.view.ShiftView;
+import org.optaweb.employeerostering.domain.spot.Spot;
+
+public final class ShiftRosterXlsxFileIO {
+
+    public static final DateTimeFormatter DATE_TIME_FORMATTER
+            = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm", Locale.ENGLISH);
+
+    // Cannot build instances of this, so private empty constructor
+    private ShiftRosterXlsxFileIO() {
+    }
+
+    public static byte[] getExcelBytesForShiftRoster(ShiftRosterView shiftRosterView) throws IOException {
+        try (Workbook workbook = new XSSFWorkbook()) {
+            Map<Long, String> employeeIdToNameMap = shiftRosterView.getEmployeeList().stream()
+                    .collect(Collectors.toMap(Employee::getId, Employee::getName));
+            for (Spot spot : shiftRosterView.getSpotList()) {
+                Sheet sheet = workbook.createSheet(spot.getName());
+                Row headerRow = sheet.createRow(0);
+                Cell headerCell = headerRow.createCell(0);
+                headerCell.setCellValue("Start");
+
+                headerCell = headerRow.createCell(1);
+                headerCell.setCellValue("End");
+
+                headerCell = headerRow.createCell(2);
+                headerCell.setCellValue("Employee");
+
+                int rowNumber = 1;
+                LocalDateTime lastShiftStartDateTime = null;
+                LocalDateTime lastShiftEndDateTime = null;
+                for (ShiftView shift : shiftRosterView.getSpotIdToShiftViewListMap().get(spot.getId())) {
+                    Row shiftRow = sheet.createRow(rowNumber);
+                    if (!shift.getStartDateTime().equals(lastShiftStartDateTime) ||
+                            !shift.getEndDateTime().equals(lastShiftEndDateTime)) {
```

Comment:
This deserves a test.

Reply:
Hint: writing tests forces us to write testable code.

In this case, the method `getExcelBytesForShiftRoster()` needs to break down into A) _roster model_ -> _excel model (workbook)_, B) _workbook_ -> _bytes_.

Reply:
I don't understand your hint. It is easy to check the workbook contents by reading the bytes in a workbook, and functions calling the method expect the bytes representing the workbook not the workbook itself. Additionally, unlike A), B) is extremely simple.

Reply:
The hint was a shortcut for "tests imply testable code" and "testable code is better than code that is hard to test".

The point is that a code that is testable usually also exhibits some good design qualities, for example Single Responsibility Principle. Writing tests guarantees testable code. Therefore writing test sometimes forces us to improve the code design simply out of necessity.

I don't think that it is easy to verify the workbook content by reading bytes. At least it's not necessary. This code is not concerned about how the workbook is represented in bytes. That's Apache POI's business. Our concern are the sheets, rows and columns in the workbook.

The second part of my previous comment suggests to break the method down into smaller pieces, specifically:
```java
public static Workbook workbook(ShiftRosterView shiftRosterView) throws IOException { ... }
public static byte[] bytes(Workbook workbook) throws IOException { ... }
```
If you do this, you can:
1. Have a test that focuses on the workbook content without jumping through hoops (reading bytes):
    ```java
    // given
    ShiftRosterView shiftRosterView = new ...;
    // when
    Workbook workbook = ShiftRosterXlsxFileIO.workbook(shiftRosterView);
    // then
    assertThat(workbook.getNumberOfSheets()).isEqualTo(...);
    // deeper inspection...
    ```
2. You can even test the second part of the story:
    ```java
    // given
    Workbook workbook = new ...;
    // when
    byte[] bytes = ShiftRosterXlsxFileIO.bytes(workbook);
    // then
    // ... (convert bytes to an input stream) ...
    // verify reading the output bytes gives the original workbook
    assertThat(new Workbook(inputStream)).isEqualTo(workbook);
    ```

All of this without going through the controller and initializing Spring context and web environment.

> ...and functions calling the method expect the bytes representing the workbook not the workbook itself.

And, of course, you can keep the original method and let it compose the smaller pieces:
```java
public static byte[] getExcelBytesForShiftRoster(ShiftRosterView shiftRosterView) throws IOException {
    return bytes(workbook(shiftRosterView));
}
```